### PR TITLE
Use the same settings UI as the devtools perf panel

### PR DESCRIPTION
**Pull request by past**
This PR copies over the new devtools performance panel settings style to the add-on. I literally copied over the generated markup and tweaked the JS event handling and initialization to match the simpler add-on architecture. I tried to keep things sufficiently generic, so there is handling for the 'java' profiler feature, even though the add-on doesn't currently work on Fennec (#33). 

I also took the liberty to remove the Information widget, which I believe we no longer consider useful. I inlined some colors that were using CSS variables in the devtools panel, since we don't yet support themes in the add-on.

### CODE REVIEW

**Comment by gregtatum** [comment]
I didn't get to this today, I'll try and look at it on Monday.

**Comment by gregtatum** [comment]
UI feedback:

* I don't think we should have 2 levels of collapsed information. In DevTools the settings are open by default, but the threads and features are collapsed. So I think we should either remove the first or second level of hiding.

* The "Apply" button drops off the screen when expanding things fully open, which I think it should always be visible. I think a viable solution would be to anchor the button at the bottom, and a nice (optional) UX feature would be to only show the apply button once a setting is changed. image

* The width feels a little cramped, especially with the two column feature labels. I feel like we should expand it to something around 400-450px.

* Expanding the features or threads menu causes an awkward animation transition where the popup window expands in size, but the inner contents do not until a second later: image

* The threads and features showing and hiding feature is causing my computer to play the error ding. I'm not sure why, and it feels a bit intermittent.

In summary, my recommendation would be to remove the second level indent; to have the threads and features non-collapsible and always open. Also when I was playing with this idea locally, there is some styling that is making the spacing between the dropdown labels inconsistent. Note that this issue exists in the DevTools UI, but feels more blatant in the tighter UI space of the pop-up.

File: popup.css [hunk 1]
```diff
@@ -255,20 +255,6 @@ kbd {
   background-color: hsla(0, 90%, 40%, 0.5);
 }
 
-.relevancy-level {
```
Comment (gregtatum) [inline popup.css:258]: @mstange Are you OK with the relevancy level going away? It seems fine to me. I don't really recall the decision to do this in the DevTools panel.

image

File: background.js [hunk 2]
```diff
@@ -164,27 +155,39 @@ async function restartProfiler() {
   if (!window.profilerState) {
     window.profilerState = {};
 
-    // Screenshots are currently only working on mac.
-    let enableScreenshots = false;
-    let platform = await browser.runtime.getPlatformInfo();
-    if (platform.os === 'mac') {
-      enableScreenshots = true;
+    const features = {
```
Comment (gregtatum) [inline background.js:158]: This is already a fairly large UI change, which I'm more concerned about than in other projects due to the lack of any testing infrastructure in place on the add-on. This is feature change feels out of scope, and will make regression tracking harder. Would you mind breaking it out into a separate PR?

File: popup.js [hunk 3]
```diff
@@ -204,38 +203,85 @@ document
  * This helper initializes and adds listeners to the features checkboxes that
  * will adjust the profiler state when changed.
  */
-async function setupFeatureCheckbox(featureName) {
+async function setupFeatureCheckbox(name) {
```
Comment (gregtatum) [inline popup.js:206]: This first part seems like it should be broken out with the other "feature" code.

File: popup.js [hunk 4]
```diff
@@ -204,38 +203,85 @@ document
...
+      // Split on commas
+      .split(',')
+      // Clean up any extraneous whitespace
+      .map(string => string.trim())
+      // Filter out any blank strings
+      .filter(string => string)
+  );
+}
+
+for (let name of features) {
```
Comment (gregtatum) [inline popup.js]: Optional: I'm a little concerned with imperative code being run intermixed with function definitions. However, there are many query selectors around doing something similar around this code. These `for` loops stand out as a bit more egregious. Typically I prefer a setup/init function to make the order of operations a bit more explicit and clear. Really though, cleaning this up I think would require a separate PR.

File: popup.js [hunk 5]
```diff
@@ -34,15 +63,18 @@ function renderControls(state) {
...
-  document.querySelector('.responsiveness-checkbox').value =
-    state.responsiveness;
-  document.querySelector('.trackopts-checkbox').value = state.trackopts;
-  document.querySelector('.js-checkbox').value = state.js;
-  document.querySelector('.screenshots-checkbox').value = state.screenshots;
-  document.querySelector('.seqstyle-checkbox').value = state.seqstyle;
-  document.querySelector('.tasktracer-checkbox').value = state.tasktracer;
-  document.querySelector('.threads-textbox').value = state.threads;
+
+  for (let name of features) {
```
Comment (gregtatum) [inline popup.js:67]: Optional: My preference is to `const` everything that can be, although we have no linting here to enforce it. In this for loop `name` is never rebound.

```js
for (const name of features) { ... }
```

**Review by gregtatum — CHANGES_REQUESTED** [review]
Thanks for the PR. Since our add-on doesn't have any testing infrastructure I'm more nervous about code changes. In summary the things I think should be handled before landing:

* Addressing UI feedback, I have some recommendations, but these are open to discussion.
* Breaking out the "feature" changes to a separate PR.
* Getting feedback on removing the relevancy feature.
* Running the HTML through Prettier.

File: popup.css (hunk 1 above)
Comment (mstange) [reply popup.css:258]: Yes that seems fine. It was a cute but rather useless gimmick.

File: background.js (hunk 2 above)
Comment (past) [reply background.js:158]: I'm not sure I understand which part of these changes you find concerning. The bulk of the changes in this file is for the added checkboxes that require initializing of their corresponding feature controls, so that's probably not what you meant. 

Another change is moving the 'leaf' feature initialization along with the rest, because currently it's very easy to miss and assume that it is uninitialized. That's just 2 lines of code, so again, probably not what you meant.

The last change is about making the "java" feature conditional on the platform instead of always disabled by default. I admit that this is a bit egregious, given that the add-on doesn't support Fennec at the moment, but it seems simple enough, generalizes the platform-dependent initialization code, and gives us less work for #33. Is this part what you'd like me to move to a separate PR?

File: popup.js (hunk 3 above)
Comment (past) [reply popup.js:206]: Which "first part" are you referring to? The rename of `featureName` to `name` or some of the code earlier in this file?

If you mean the argument renaming, I can certainly take it out, but it seemed a bit more uniform with `setupThreadCheckbox(name)`, even though I didn't end up combining them as I was initially planning to do.

File: background.js (hunk 2 above)
Comment (gregtatum) [reply background.js:158]: Coming back to this, I think I overlooked that the following snippet was in the original code.

```
if (platform.os === 'mac') {
      enableScreenshots = true;
}
```

There is still some behavior change here, but it's less than I thought. I'd be fine if you kept it in this PR.

> I'm not sure I understand which part of these changes you find concerning

Also, some of my comments were on the state of the codebase, not specifically on your code changes.

File: popup.js (hunk 3 above)
Comment (gregtatum) [reply popup.js:206]: I was certainly not communicating well on this PR. I believe I was referring to the:

```

  // Java profiling is only meaningful on android.
  if (name == 'java') {
    if (platform.os !== 'android') {
      document.querySelector('#java').style.display = 'none';
      return;
    }
  } 
```

I changed my thoughts on it, please keep it.

File: popup.js (hunk 5 above)
Comment (past) [reply popup.js:67]: Done.

File: popup.js (hunk 4 above)
Comment (past) [reply popup.js]: I've left this for a followup PR.

**Comment by past** [comment]
I believe I have addressed all of your feedback, with the exception of the two optional change requests (displaying the 'apply' button only after changes, and moving popup initialization to a function). I'll try to look into those in followups as time permits.

**Review by julienw — CHANGES_REQUESTED** [review]
hey @past, thanks !
I tried the latest version of your patch, and I believe there are still UI-related issues, at least on Linux. I didn't look at the code because I believe Greg already did this and you fixed the comments he left.

1. there's an horizontal scrollbar:
addon-settings-1

2. the space between the columns in the `feature` section isn't big enough, see:
addon-settings-2
But because we have such a small width, I believe we shouldn't have this column layout here. Instead we should have one line with the title, and its explanation below, in a 1-column layout. I know this was something that you discussed with Greg before he left for PTO, so I'm sorry to comment about this subject again, but I think this is important here.

**Review by gregtatum — APPROVED** [review]
Looks good on mac, r+ with Julien's Linux feedback. Thanks for addressing the UI issues on it.

One other UX follow-up would be to close the settings menu after the user applies the changes. It feels like I don't really want it open anymore, and I'm not sure my changes have been applied.

**Comment by past** [comment]
I've been trying to find some time to work on this, but it doesn't look like I will be able to do so any time soon. If anyone else has any cycles to take this over the finish line, I would appreciate it.

**Comment by canova** [comment]
I can take this over since I also need to write a patch that depends on this(for the fixed-time-window profiling mode).

**Comment by canova** [comment]
Addressed the review comments. Now features side looks like this:

screen shot 2018-10-22 at 6 54 23 pm

@julienw Could you look at the patch and also check if the checkbox and the description alignment is also like this on your side?

**Comment by julienw** [comment]
There's still a problem with the feature labels, see the screenshot:
image

The problem is the `width` in https://github.com/devtools-html/Gecko-Profiler-Addon/pull/125/files#diff-b5dfc3ff7c7bccd1bec16cd3bc5f663eR388 that's now useless:
```css
.perf-settings-feature-name {
  width: 100px; // <-- you can remove this
  color: #0060df;
  line-height: 1.6;
}
```

**Review by julienw — APPROVED** [review]
r=me with the removal of `width: 100px`

File: popup.css [hunk 6]
```diff
@@ -12,7 +12,7 @@ html {
 
 body {
   margin: 0;
-  width: 400px;
+  max-width: 32em;
```
Comment (julienw) [inline popup.css:23]: ah, and don't forget to file a bug in Firefox about that.
Then add a comment explaining that if we use `width` instead of `max-width` we have an horizontal screenshot in Firefox, with the bug #.
